### PR TITLE
Fix stock movement creation from new order

### LINKED ISSUE #27627: Stock movement without order reference is created by undelivered new order
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

New order created from BO always add a new stock movement:
- from any order status, not only "Delivered"
- without any reference to order (id_order = NULL)

Later when the order status is switched to a status with the flag "shipped" activated, the stock movement is duplicated with the correct order reference

### Expected behavior

- Stock movement should be created only from a new order with "delivered" status
- This order should be referenced in the newly created stock movement

Initial specifications: https://forge.prestashop.com/browse/FF-1

### Steps to reproduce

1. Create a new order from BO with "Waiting for payment" status:
2. Observe the new stock movement inserted in database table ps_stock_mvt with id_order = NULL

### PrestaShop version(s) where the bug happened

8.0, 1.7.8.x

### PHP version(s) where the bug happened

7.4.3

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by saulaski**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | New order created from BO always add a new stock movement:<ul><li>from any order status, not only "delivered"</li><li>without any reference to order (id_order = NULL)</li></ul>This PR fixes the bug:<ul><li>stock movement is only created from an order with "delivered" status</li><li>this order is referenced in the newly created stock movement</li></ul>As required by initial specifications: https://forge.prestashop.com/browse/FF-1
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | <ul><li>Method `OrderDetail::checkProductStock()` has been renamed to `OrderDetail::updateProductQuantityInStock()`</li><li>Protected properties have been removed in abstract class `PrestaShopBundle\Entity\Repository\StockManagementRepository`:<ul><li>$languageId</li><li>$shopId</li><li>$context</li></ul>Use new methods instead:<ul><li>`getCurrentContext(): Context` Returns the current context</li><li>`getCurrentEmployee(): Employee` Returns the employee set in current context</li><li>`getCurrentLanguageId(): int` Returns the language ID of the employee in current context</li><li>`getCurrentShop(): Shop` Returns the shop set in current context</li><li>`getContextualShopId(): int` Returns the contextual ID of the shop in current context</li></ul></li></ul>
| Deprecations?     | no
| Fixed ticket?     | Fixes #27627.
| How to test?      | <ol><li>Create a new order from BO with "Waiting for payment" status</li><li>No stock movement has been inserted in database table ps_stock_mvt</li><li>Create a new order from BO with 'Shipped" or "Delivered" status</li><li>A new stock movement has been inserted in database table ps_stock_mvt with id_order != NULL</li></ol>
| Possible impacts? | Any feature related to stock management update by orders.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27626)

### CODE REVIEW

**Comment by saulaski** [comment]
This bug fix might cause BC breaks in previous versions 1.7, in case stock management relies on invalid movements.
Should we add the fix in 1.7.8.x anyway ?

**Review by matthieu-rolland — COMMENTED** [review]
Seems like behat tests are not happy with the bug fix...  not sure if it's the tests that need to be adapted or if the problem lies in the fix 🤔

**Comment by saulaski** [comment]
> Seems like behat tests are not happy with the bug fix... not sure if it's the tests that need to be adapted or if the problem lies in the fix thinking

Behat tests are broken indeed, and the issue is indirectly related to this bug fix. This fix persists stock movements in database only when required by business rules (order status = delivered). It means that stock movement repository is now instanciated from Symfony container only when it is required by order status.

This repository throws an exception if there is no single shop in current context (see error message in tests). A few Behat tests don't setup the context correctly, but nobody could notice this edge case because the repository was already available in Symfony container.
The repository was always instanciated by previous tests to persist stock movements, even if they shouldn't according to business rules.

File: classes/order/OrderDetail.php [hunk 1]
```diff
@@ -536,25 +536,38 @@ protected function setVirtualProductInformation($product)
      * Check the order status.
```
Comment (eternoendless) [inline classes/order/OrderDetail.php]: Perhaps we can take this opportunity to update this description?

File: classes/order/OrderDetail.php [hunk 2]
```diff
@@ -536,25 +536,38 @@ protected function setVirtualProductInformation($product)
...
+
+        if (!StockAvailable::dependsOnStock($product['id_product'])) {
+            $quantityUpdated = StockAvailable::updateQuantity(
+                $product['id_product'],
+                $product['id_product_attribute'],
+                -(int) $product['cart_quantity'],
+                $product['id_shop'],
+                $orderStateId == Configuration::get('PS_OS_DELIVERED'),
+                [
+                    'id_order' => $this->id_order,
```
Comment (eternoendless) [inline classes/order/OrderDetail.php:560]: The stock movement type is missing from the parameters in this method call. Logically, any stock movement should include an `id_stock_mvt_reason` (see https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Stock/StockManager.php#L396-L398) so that the merchant can later tell why the stock was changed.

However, there is no stock movement type for "order shipped" as we can see in ps_configuration:
Screenshot 2022-02-09 at 16 45 54

It's not clear to me whether this was forgotten or done in purpose. According to the original spec, tracking physical and available quantities separately is out of scope, so following this logic (we only track available quantities), the available stock should be decremented when the order is created, not when the order is shipped. This would be consistent with the "order shipped" stock movement type not existing.

Please advise @MatShir @fabienvallon

Comment (saulaski) [reply classes/order/OrderDetail.php:560]: @eternoendless Agreed.

I got it wrong when I assumed the default movement reason was `PS_STOCK_MVT_DEC_EMPLOYEE_EDITION`, but it's a quick fix.
A point remains unclear. An order created from BO should get treated as an usual customer order, or as an employee edition ?
I suppose the "Customer Order" reason includes the "order shipped" state, which has no impact on any quantity type.

Unfortunately the current behavior in FO is inconsistent with physical and available quantity tracking:
- a new order adds no stock movement until delivered
- movement is created only when order status is switched to delivered
- movement reason ID is 3 => `PS_STOCK_MVT_REASON_DEFAULT` or `PS_STOCK_CUSTOMER_ORDER_REASON`

And the current behavior in BO (before this PR) is messed up as well:
- a new order adds a new movement, whatever the order status, without order reference
- switching order status to "Delivered" duplicates the same movement with correct order reference

As you mentioned, tracking available quantity requires a new movement immediately when the order is created.

Now should we keep FO behaviour ? or fix both FO and BO to track available quantity accurately ?
Tracking available quantity should only handle 2 events :
- New order with any status
- Product return

Comment (saulaski) [reply classes/order/OrderDetail.php:560]: @eternoendless IMHO the easiest way to track both available and physical quantities is to split customer order reason in 2 reasons :
- Customer Order Reserve => impacts available quantity only when order is created
- Customer Order Shipping => impacts physical quantity only when order status is updated to shipping

Optionally it would make sense to distinguish orders created by an employee in BO: 
- Employee Order Reserve => available quantity
- Employee Order Shipping => physical quantity

Of course relying on movement reasons only is not the safest design, and stock movements might need a new `type` property:
- `reserve` type => only from order creation
- `shipping` type => only from the 1st order status set to "Shipping" or "Delivered" (order workflow is more complex to handle)
- `physical` type => everything else

Then sum all movement quantities:
- available tracking should only include both `reserve` and `physical` movements
- physical tracking should only include both `shipping` and `physical` movements

File: classes/order/OrderDetail.php (hunk 1 above)
Comment (saulaski) [reply classes/order/OrderDetail.php]: Indeed, descriptions have been updated:
- PhpDoc: `Updates product quantity in stock, according to order status.`
- Method name: `updateProductQuantityInStock()`

File: classes/order/OrderDetail.php [hunk 3]
```diff
@@ -533,28 +533,42 @@ protected function setVirtualProductInformation($product)
...
-            }
-            Product::updateDefaultAttribute($product['id_product']);
+        $dismissOrderStateIds = Configuration::getMultiple([
+            'PS_OS_CANCELED',
+            'PS_OS_ERROR',
+        ]);
+        if (in_array($orderStateId, $dismissOrderStateIds)) {
+            return;
+        }
+        $quantityUpdated = true;
```
Comment (Amoifr) [inline classes/order/OrderDetail.php]:
```suggestion
        $isQuantityUpdated = true;
```

File: classes/order/OrderDetail.php [hunk 4]
```diff
@@ -533,28 +533,42 @@ protected function setVirtualProductInformation($product)
...
+
+        if (!StockAvailable::dependsOnStock($product['id_product'])) {
+            $quantityUpdated = StockAvailable::updateQuantity(
+                $product['id_product'],
+                $product['id_product_attribute'],
+                -(int) $product['cart_quantity'],
+                $product['id_shop'],
+                $orderStateId == Configuration::get('PS_OS_DELIVERED'),
```
Comment (Amoifr) [inline classes/order/OrderDetail.php]: not a strict compare with === ? (type can differ ?)

File: classes/order/OrderDetail.php [hunk 5]
```diff
@@ -533,28 +533,42 @@ protected function setVirtualProductInformation($product)
...
+                [
+                    'id_order' => $this->id_order,
+                    'id_stock_mvt_reason' => Configuration::get('PS_STOCK_CUSTOMER_ORDER_REASON'),
+                ]
+            );
+        }
+        if ($quantityUpdated) {
```
Comment (Amoifr) [inline classes/order/OrderDetail.php]:
```suggestion
        if ($quantityUpdated === true) {
```

File: classes/order/OrderDetail.php [hunk 6]
```diff
@@ -533,28 +533,42 @@ protected function setVirtualProductInformation($product)
...
+            $product['stock_quantity'] -= $product['cart_quantity'];
```
Comment (Amoifr) [inline classes/order/OrderDetail.php:569]: also if $product['stock_quantity'] become < 0 ?

Comment (Amoifr) [reply classes/order/OrderDetail.php:569]: allow negative stock ?

File: src/PrestaShopBundle/Entity/Repository/StockRepository.php [hunk 7]
```diff
@@ -129,7 +129,7 @@ public function updateStock(Movement $movement, $syncStock = true)
                     $product,
                     $productIdentity->getCombinationId(),
                     $delta,
-                    $this->contextAdapter->getContext()->shop->id,
+                    $this->getCurrentShop()->id,
```
Comment (Amoifr) [inline src/PrestaShopBundle/Entity/Repository/StockRepository.php:132]: different as getCurrentShopId() ?

File: classes/order/OrderDetail.php (hunk 5 above)
Comment (saulaski) [reply classes/order/OrderDetail.php]: If renamed to `$isQuantityUpdated`, the variable type looks obviously boolean. Is it absolutely required to add a test condition here ?

File: classes/order/OrderDetail.php (hunk 6 above)
Comment (saulaski) [reply classes/order/OrderDetail.php:569]: Counterintuitive I agree. I assumed the negative quantity case is allowed by the next statement down below:
```
        if ($product['stock_quantity'] < 0 && Configuration::get('PS_STOCK_MANAGEMENT')) {
            $this->outOfStock = true;
        }
```

File: src/PrestaShopBundle/Entity/Repository/StockRepository.php (hunk 7 above)
Comment (saulaski) [reply src/PrestaShopBundle/Entity/Repository/StockRepository.php:132]: Possibly yes, and it kinda bothers me :)
- current shop instance is stored in context, and its ID property is required here
- Shop class also owns a static property `$context_id_shop` which is returned by `$shop->getContextualShopId()`. This static value is refreshed by `Shop::intialize()`... very confusing indeed.

I kept both values rigorously splitted to prevent any risk of edge case. But I rename method `StockRepository::getCurrentShopId()` to `getCurrentShopId()`, because it's only a shortcut to `$shop->getContextualShopId()`.

File: classes/order/OrderDetail.php (hunk 4 above)
Comment (saulaski) [reply classes/order/OrderDetail.php]: String value from configuration could get casted to integer. But `$orderStateId` argument could be trusted without any type hint ?

File: classes/order/OrderDetail.php (hunk 3 above)
Comment (saulaski) [reply classes/order/OrderDetail.php]: Agreed and renamed, thank you

File: src/PrestaShopBundle/Entity/Repository/StockRepository.php (hunk 7 above)
Comment (Amoifr) [reply src/PrestaShopBundle/Entity/Repository/StockRepository.php:132]: magic !

File: classes/order/OrderDetail.php (hunk 6 above)
Comment (Amoifr) [reply classes/order/OrderDetail.php:569]: so go on !

File: classes/order/OrderDetail.php (hunk 5 above)
Comment (Amoifr) [reply classes/order/OrderDetail.php]: this adds a strong additional typing. To be sure that we don't allow 1 or "1" or "0" or anything else than true.

File: classes/order/OrderDetail.php (hunk 4 above)
Comment (Amoifr) [reply classes/order/OrderDetail.php]: as you pleased

Comment (saulaski) [reply classes/order/OrderDetail.php]: After second thoughts, I agree with you: strict code is safer

File: classes/order/OrderDetail.php (hunk 5 above)
Comment (saulaski) [reply classes/order/OrderDetail.php]: After second thoughts, I agree with you: strict code is safer

File: src/PrestaShopBundle/Entity/Repository/StockRepository.php (hunk 7 above)
Comment (saulaski) [reply src/PrestaShopBundle/Entity/Repository/StockRepository.php:132]: black magic :)

**Comment by saulaski** [comment]
@eternoendless @MatShir @fabienvallon
This PR is only supposed to fix the incorrect BO behavior, and make it work just like FO. Because most orders are created from FO, its behavior should prime to avoid BC breaks and data inconsistency.
This bug fix is also required to unblock another PR : https://github.com/PrestaShop/PrestaShop/pull/27407

If the current implementation in both BO and FO doesn't match the initial specifications : https://forge.prestashop.com/browse/FF-1
Then we should create a dedicated issue in order to plan the best solution and fix the behavior as initially expected.

File: classes/order/OrderDetail.php (hunk 2 above)
Comment (saulaski) [reply classes/order/OrderDetail.php:560]: Actually the technical solution I provided is out of the scope described in original specifications : https://forge.prestashop.com/browse/FF-1

A stock movement is supposed to track physical quantities only, i.e. stock quantities physically present in a given warehouse. A movement was never intended to track stock quantities available to order.

Even if the solution I provided could work in theory, it's certainly over complicated to implement 2 distinct business needs at the same technical place.
Then I propose to split the technical implementation in 2 : 
- keep stock movements to track physical quantities only
- add a new table to track available quantities only : `stock_reserve`

Then the physical quantity is easy to obtain from the stock movements table alone.
And the available quantity requires to query the stock table joined to both tables : movements & reserves

IMHO opinion this implementaiton is safer, and it allows all kind of business behaviors according to shop configuration.
Example : to deactivate the available quantity treshold, don't join stock reserves or don't insert any reserved quantity at all.

**Comment by saulaski** [comment]
PM approved, BO should behave like FO.
A new order should add a stock movement only if the product is physically "shipped", i.e. if the order status as the flag "Shipped" activated.

File: classes/order/OrderDetail.php (hunk 2 above)
Comment (jolelievre) [reply classes/order/OrderDetail.php:560]: We need to discuss this, but I believe `StockAvailable.reserved_quantity` is already here for that purpose

File: src/PrestaShopBundle/Entity/Repository/StockManagementRepository.php [hunk 8]
```diff
@@ -123,26 +109,46 @@ public function __construct(
         $this->contextAdapter = $contextAdapter;
         $this->imageManager = $imageManager;
         $this->tablePrefix = $tablePrefix;
+    }
 
-        $this->context = $contextAdapter->getContext();
+    protected function getCurrentContext(): Context
```
Comment (jolelievre) [inline src/PrestaShopBundle/Entity/Repository/StockManagementRepository.php:117]: All this is really nice!!

File: src/PrestaShopBundle/Entity/Repository/StockRepository.php (hunk 7 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Entity/Repository/StockRepository.php:132]: Yeah it's safer to do this Both values should, theoretically, be synced But in case they are not let's keep the initial behaviour for now

**Review by jolelievre — COMMENTED** [review]
I agree with all the modifications in the manager to perform Just In Time checking
The modification in OrderDetail is a bit more complicated and needs some discussions I think

File: classes/order/OrderDetail.php (hunk 2 above)
Comment (saulaski) [reply classes/order/OrderDetail.php:560]: @jolelievre 
I suppose you refer the table column as `StockAvailable.reserved_quantity` ? class `StockAvailable` owns no `reserved_quantity` property.
Looks like indeed that this column intents to keep track of the total quantity of reserved products in stock. Which is not exactly the same as historization tracking feature alike stock movements. If the product doesn't need to historize reservations and their cancellation, sounds good enough to me :)

File: src/PrestaShopBundle/Entity/Repository/StockRepository.php (hunk 7 above)
Comment (saulaski) [reply src/PrestaShopBundle/Entity/Repository/StockRepository.php:132]: [Paranoid mode enabled] ;)

**Comment by saulaski** [comment]
@MatShir @jolelievre Added new integration tests to validate stock movements according to status of the newly created order

File: tests/Integration/Behaviour/Features/Scenario/Order/order_stock.feature [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    And I add 2 products "product<index>" to the cart "dummy_cart<index>"
+    And I add order "bo_order<index>" with the following details:
+      | cart                | dummy_cart<index> |
+      | message             | test<index>       |
+      | payment module name | dummy_payment     |
+      | status              | <order_status>    |
+    Then product "product<index>" last employees stock movements should be:
+      | first_name | last_name | delta_quantity |
+      | Puff       | Daddy     | 100            |
+    Examples:
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Order/order_stock.feature:33]: Is this a feature similar to data providers in PHPUnit? I didn't know about this, it's awesome!

**Review by jolelievre — DISMISSED** [review]
Thanks @saulaski

File: tests/Integration/Behaviour/Features/Scenario/Order/order_stock.feature (hunk 9 above)
Comment (saulaski) [reply tests/Integration/Behaviour/Features/Scenario/Order/order_stock.feature:33]: I didn't know either until yesterday, because I am too lazy to duplicate the same scenario.
And indeed I found another solution similar to data providers in Behat documentation :
https://behat.org/en/latest/user_guide/writing_scenarios.html#scenario-outlines

File: classes/order/OrderDetail.php [hunk 10]
```diff
@@ -533,28 +533,46 @@ protected function setVirtualProductInformation($product)
...
     /**
-     * Check the order status.
+     * Updates product quantity in stock, according to order status.
      *
      * @param array $product
-     * @param int $id_order_state
+     * @param int $orderStateId
      */
-    protected function checkProductStock($product, $id_order_state)
+    protected function updateProductQuantityInStock($product, $orderStateId)
```
Comment (Progi1984) [inline classes/order/OrderDetail.php]: Could you typehint the new function, please ?

Comment (Progi1984) [reply classes/order/OrderDetail.php]: Could you document the BC break for the removal of the old method  ?

File: src/PrestaShopBundle/Entity/Repository/StockManagementRepository.php [hunk 11]
```diff
@@ -75,21 +76,6 @@ abstract class StockManagementRepository
...
-    /**
-     * @var int
-     */
-    protected $shopId;
-
-    /**
-     * @var \Context
-     */
-    protected $context;
-
```
Comment (Progi1984) [inline src/PrestaShopBundle/Entity/Repository/StockManagementRepository.php:92]: Could you document the BC break for the removal of these attributes  ?

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks (about BC Break) but the base is good.

File: classes/order/OrderDetail.php (hunk 10 above)
Comment (saulaski) [reply classes/order/OrderDetail.php]: Is it really a BC break, since the method is protected ? BC would break only if the class is extended outside of core components.

File: src/PrestaShopBundle/Entity/Repository/StockManagementRepository.php (hunk 11 above)
Comment (saulaski) [reply src/PrestaShopBundle/Entity/Repository/StockManagementRepository.php:92]: Same... BC would break only if the class is extended outside of core components.

File: classes/order/OrderDetail.php (hunk 10 above)
Comment (Progi1984) [reply classes/order/OrderDetail.php]: Yes : the class is not final and it's a protected. So people would have been able to override this class. (And yes it's possible to extend legacy).

File: src/PrestaShopBundle/Entity/Repository/StockManagementRepository.php (hunk 11 above)
Comment (Progi1984) [reply src/PrestaShopBundle/Entity/Repository/StockManagementRepository.php:92]: It's possible to extend legacy

**Comment by saulaski** [comment]
> Some feedbacks (about BC Break) but the base is good.

BC breaks are now properly documented

**Review by jolelievre — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Comment by hibatallahAouadni** [comment]
Hello @saulaski 

Thanks for your PR :rocket: 
From a fresh install, we use to have some lines in `PREFIX_stock_mvt` of the created Orders (by default) while no one of them has the `Delivered` or `Shipped` Status:

image

- With your PR(fresh install), I don't have any line in the table :clap: :heavy_check_mark: 
- When I create an Order with any status but `Delivered` or `Shipped`, no new line hasn't insert :heavy_check_mark: 
- Creating an Order with a status `Delivered` or `Shipped`, a new line will be add in the `PREFIX_stock_mvt` table with `id_order` :heavy_check_mark: 

image

image

But I have a question, the specs says:
> **A stock movement is a "log" of any physical stock modification**. It allows a precise tracking. For instance when a customer order is shipped, it creates a movement that indicates the number of products removed from physical stock, the date, time and the reason.

IMHO, I think Physical stock modification doesn't concern a Virtual product :thinking: 
Cause when I created an Order with only one Virtual product and with `Delivered` Status, a new line will be created in the table `PREFIX_stock_mvt`, and I know it's confusing to change the status to `Delivered` or `Shipped` for a virtual Order :sweat_smile: 
So ping @PrestaShop/product-team, what should we do? we have two solutions:

1. add a new feature: when we have a virtual order (an order only contains virtual products), there is no `Delivered` or `Shipped` status in the dropdown
2. when we change virtual order status to `Delivered` or `Shipped`, no new row in `PREFIX_stock_mvt` table will be created.

Waiting for your feedback.
Thanks!

**Comment by saulaski** [comment]
@hibatallahAouadni Thanks a lot for your testing :)

This issue has been discussed a lot with a PM, and we haven't mentioned the virtual product case indeed.

@PrestaShop/product-team should we exclude stock movements of virtual products when a new order is created ? what is the business rule to apply here ?

**Comment by saulaski** [comment]
@hibatallahAouadni added a condition to dismiss stock movements for virtual products

The simplest solution since I got no news from @PrestaShop/product-team

**Review by jolelievre — DISMISSED** [review]
It raises the question about stock movement when employee performs an update But probably not the scope of this PR

It might be worth to homogenize this behaviour everywhere for virtual product though

@PrestaShop/product-team what do you think?

**Comment by hibatallahAouadni** [comment]
Ping @MatShir friendly reminder :pray:

**Comment by MatShir** [comment]
Thanks @hibatallahAouadni for your suggestion, but I think just blocking the stock physical mvt if enough. Maybe to avoid confusion, I'm wondering if we should rename the stock mvt page to like "physical mvt"

File: classes/order/OrderDetail.php [hunk 12]
```diff
@@ -533,28 +533,45 @@ protected function setVirtualProductInformation($product)
...
+                    'id_stock_mvt_reason' => Configuration::get('PS_STOCK_CUSTOMER_ORDER_REASON'),
+                ]
+            );
+        } else {
+            $isQuantityUpdated = true;
+        }
+        if ($isQuantityUpdated === true) {
+            $product['stock_quantity'] -= $product['cart_quantity'];
+        }
+        if ($product['stock_quantity'] < 0 && Configuration::get('PS_STOCK_MANAGEMENT')) {
```
Comment (atomiix) [inline classes/order/OrderDetail.php:571]: I'm maybe missing something (and out of scope) but shouldn't the product be out of stock when `stock_quantity <= 0` instead of `< 0`?

Comment (jolelievre) [reply classes/order/OrderDetail.php:571]: Good call, but I think it's ok this way If the remaining stock is strictly 0 it means the order just got the last left items 

I believe the `$this->outOfStock` parameter is here to indicate that this order is explicitly using some nonexisting stock (probably linked the the availability to order Out of stock products)

**Review by atomiix — DISMISSED** [review]

**Comment by hibatallahAouadni** [comment]
Hello @saulaski 

Thanks again for your PR!
## From fresh Install :heavy_check_mark: 

1. I don't have any line in the table :clap: :heavy_check_mark: 
2. When I create an Order with any status but `Delivered` or `Shipped`, no new line hasn't insert :heavy_check_mark: 

## Create an Order from BO 

1. **Simple product:** a line added to `PREFIX_stock_mvt` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
3. **Product with combination:** a line added to `PREFIX_stock_mvt` only  when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
4. **Product Customizable:** a line added to `PREFIX_stock_mvt` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
5. **Pack of Products:** a line added to `PREFIX_stock_mvt` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
6. **Pack of Products (contains a Virtual Product):** a line added to `PREFIX_stock_mvt` without **`id_Order`** and the order status is not `Delivered` nor `Shipped` :x: 
7. **Virtual Product:** **`no line`** added to `PREFIX_stock_mvt` when the status of the created order is `Delivered` or `Shipped` :heavy_check_mark: 
8. **Virtual Product:** a line added to `PREFIX_stock_mvt` when the status of the order use to be any other status and then changed to `Delivered` or `Shipped` :x: 
9. **Simple product & product with combinations**: two lines added to `PREFIX_stock_mvt` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
10. **Simple product & Virtual product**: **`only one line`** (of the simple product) added to `PREFIX_stock_mvt` only when the created order has the status `Delivered` or `Shipped` :heavy_check_mark: 
11. **Simple product & Virtual product**: two lines added to `PREFIX_stock_mvt` only  when the status of the order use to be any other status and then changed to `Delivered` or `Shipped` :x: 
12. **Customizable product & Virtual product**: **`only one line`** (of the Customizable product) added to `PREFIX_stock_mvt` only when the created order has the status `Delivered` or `Shipped` :heavy_check_mark: 
13. **Customizable product & Virtual product**: two lines added to `PREFIX_stock_mvt` only when the status of the order use to be any other status and then changed to `Delivered` or `Shipped` :x: 
14. **Product with combinations & Virtual product**: **`only one line`** (of the Customizable product) added to `PREFIX_stock_mvt` only when the created order has the status `Delivered` or `Shipped` :heavy_check_mark: 
15. **Product with combinations & Virtual product**: two lines added to `PREFIX_stock_mvt` only when the status of the order use to be any other status and then changed to `Delivered` or `Shipped` :x: 

## Create an Order from FO 

1. **Simple product:** a line added to `PREFIX_stock_mvt` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
3. **Product with combination:** a line added to `PREFIX_stock_mvt` only  when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
4. **Product Customizable:** a line added to `PREFIX_stock_mvt` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
5. **Pack of Products:** a line added to `PREFIX_stock_mvt` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
6. **Pack of Products (contains a Virtual Product):** a line added to `PREFIX_stock_mvt` without **`id_Order`** and the order status is not `Delivered` nor `Shipped` :x: 
7. **Virtual Product:** a line added to `PREFIX_stock_mvt` only when the status of the order changed to `Delivered` or `Shipped` :x: 
8. **Simple product & product with combinations**: two lines added to `PREFIX_stock_mvt` when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
9. **Simple product & Virtual product**: two lines added to `PREFIX_stock_mvt` when the status of the order changed to `Delivered` or `Shipped` :x: 
10. **Customizable product & Virtual product**: two lines added to `PREFIX_stock_mvt` when the status of the order changed to `Delivered` or `Shipped` :x: 
11. **Product with combinations & Virtual product**: two lines added to `PREFIX_stock_mvt` when the status of the order changed to `Delivered` or `Shipped` :x: 

## B2B & Multistore
Same issues as above

## TLDR
All the raised issues are related to Virtual Product :disappointed: 
For the Pack of Products that contain a Virtual Product: I encountred weird behavior (a new line added to the tab without id_order), check the attached screenshot below:

image

Thanks for your patience :pray: 
Waiting for your feedback.

PS: I think we may consider all behaviors related to Virtual Product are out of scope of this PR, but we should (at least) fix the weird behavior of the pack (contains Virtual product).

Thanks!

File: classes/order/OrderDetail.php (hunk 12 above)
Comment (saulaski) [reply classes/order/OrderDetail.php:571]: `outOfStock` property seems only read by payment module from `OrderDetail::getStockState()`, in order to persist order history and to send email to customer.
https://github.com/PrestaShop/PrestaShop/blob/d1d4982ee74c70ac331de2b76cb00c84912d820c/classes/PaymentModule.php#L584-L595

**Comment by saulaski** [comment]
@hibatallahAouadni Thank you for this great testing that provides a much better view on the current functional situation. This is some serious retro-engineering that will for sure help @MatShir.
Looks like the new "virtual product" rule introduced in this PR is limited by its scope: order creation from BO only.
The new rule isn't implemented yet in unrelated use cases:
- order creation from FO
- order status update to `Shipped` or `Delivered`
- pack of products which seems to be well hidden in its own and mysterious piece of code

IMHO the new "virtual product" rule should be handled by a new dedicated issue, and get implemented in all use cases at once for consistency sake.
The initial purpose of the current PR was only to fix inconsistent behaviors between BO and FO at order creation. BO added stock movement lines whereas FO didn't.

@MatShir what do you think ? should we revert the new "virtual product" rule here, and move the feature into a new issue ?

**Comment by MatShir** [comment]
Thanks a lot for your test @hibatallahAouadni, It seems there is a lot of rule that are not well-defined yet. We will use your tests for the next specs/issue to fix the overall issue.
Let's revert the changes about the virtual product behavior, they were out of the scope. Let's keep the scope to the order creation.

**Review by jolelievre — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

**Comment by saulaski** [comment]
@MatShir @jo @hibatallahAouadni New "virtual products" rule has been reverted

**Review by atomiix — APPROVED** [review]

**Review by jolelievre — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Comment by hibatallahAouadni** [comment]
Hello again,

## From fresh Install :heavy_check_mark: 

1. I don't have any line in the table :clap: :heavy_check_mark: 
2. When I create an Order with any status but `Delivered` or `Shipped`, no new line hasn't insert :heavy_check_mark: 

## Create an Order from BO :heavy_check_mark: 

1. **Simple product:** a line added to `PREFIX_stock_mvt` with `id_Order` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
3. **Product with combination:** a line added to `PREFIX_stock_mvt` with `id_Order` only  when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
4. **Product Customizable:** a line added to `PREFIX_stock_mvt` with `id_Order` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
5. **Pack of Products:** a line added to `PREFIX_stock_mvt` with `id_Order` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
6. **Simple product & product with combinations**: two lines added to `PREFIX_stock_mvt` with `id_Order` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
7. **Simple product & Product Customizable:**  two lines added to `PREFIX_stock_mvt` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
8. **Simple product & Pack of Products:** two lines added to `PREFIX_stock_mvt` with `id_Order` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
9. **Product with combination & Product Customizable:** two lines added to `PREFIX_stock_mvt` with `id_Order` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
10. **Product with combination & Pack of Products:** two lines added to `PREFIX_stock_mvt` with `id_Order` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
11. **Product Customizable & Pack of Products:** two lines added to `PREFIX_stock_mvt` with `id_Order` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
12. **Simple product & product with combinations & Product Customizable & Pack of Products:** four lines added to `PREFIX_stock_mvt` with `id_Order` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 

## Create an Order from FO :heavy_check_mark: 

1. **Simple product:** a line added to `PREFIX_stock_mvt` with `id_Order` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
3. **Product with combination:** a line added to `PREFIX_stock_mvt` with `id_Order` only  when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
4. **Product Customizable:** a line added to `PREFIX_stock_mvt` with `id_Order` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
5. **Pack of Products:** a line added to `PREFIX_stock_mvt` with `id_Order` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
6. **Simple product & product with combinations**: two lines added to `PREFIX_stock_mvt` with `id_Order` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
7. **Simple product & Product Customizable:**  two lines added to `PREFIX_stock_mvt` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
8. **Simple product & Pack of Products:** two lines added to `PREFIX_stock_mvt` with `id_Order` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
9. **Product with combination & Product Customizable:** two lines added to `PREFIX_stock_mvt` with `id_Order` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
10. **Product with combination & Pack of Products:** two lines added to `PREFIX_stock_mvt` with `id_Order` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
11. **Product Customizable & Pack of Products:** two lines added to `PREFIX_stock_mvt` with `id_Order` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 
12. **Simple product & product with combinations & Product Customizable & Pack of Products:** four lines added to `PREFIX_stock_mvt` with `id_Order` only when the status of the order changed to `Delivered` or `Shipped` :heavy_check_mark: 

## Multistore & B2B :heavy_check_mark: 

All good (same as above) :heavy_check_mark: 

I opened a new issue for all cases related to Virtual Product #27993
LGTM, QA :heavy_check_mark: 

Thanks!

**Comment by Progi1984** [comment]
Thanks @saulaski & @hibatallahAouadni
